pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+	// events since they'll be handled in the OnUpdatePod callback.
+	// GH issue #13136
+	id := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+	if _, ok := manager.policyPods[id]; ok {
+		return
+	}
+	manager.OnUpdatePodLocked(pod, false, true)
+}
+
+// OnUpdatePodLocked is the handler for pod info updates.
+func (manager *Manager) OnUpdatePodLocked(pod *slim_corev1.Pod, removeOld bool, upsertNew bool) {
+	var podData *podMetadata
+	var err error
+
+	if len(manager.policyConfigs) == 0 {
+		return
+	}
+
+	podIPs := k8sUtils.ValidIPs(pod.Status)
+	if len(podIPs) == 0 {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sPodName:   pod.Name,
+			logfields.K8sNamespace: pod.Namespace,
+		}).Error("Failed to get valid pod ips, skipping update to Egress Policy")
+		return
+	}
+
+	if podData, err = manager.getPodMetadata(pod, podIPs); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sPodName:   pod.Name,
+			logfields.K8sNamespace: pod.Namespace,
+		}).Error("Failed to get valid pod metadata, skipping update to Egress Policy")
+		return
+	}
+
+	if removeOld {
+		// Check if the pod was previously selected by any of the policies.
+		if policies, ok := manager.policyPods[podData.id]; ok {
+			for _, policy := range policies {
+				config := manager.policyConfigs[policy]
+				err := manager.deleteEgressMap(config, podData)
+				if err != nil {
+					log.WithError(err).Errorf("Error updating pod mapping for pod %v", podData.id)
+					return
+				}
+			}
+			delete(manager.policyPods, podData.id)
+			delete(manager.podDataStore, podData.id)
+		}
+	}
+
+	if upsertNew {
+		// Check if any of the current policies select this pod.
+		for _, config := range manager.policyConfigs {
+			if config.policyConfigSelectsPod(podData) {
+				err := manager.upsertPolicyPod(config, podData)
+				if err != nil {
+					log.WithError(err).Errorf("Error upserting pod mapping for pod %v", podData.id)
+					return
+				}
+			}
+		}
+	}
+}
+
+//OnUpdatePod is the event handler for pod updates.
+func (manager *Manager) OnUpdatePod(pod *slim_corev1.Pod, needsReassign bool) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	manager.OnUpdatePodLocked(pod, needsReassign, true)
+}
+
+//OnDeletePod is the event handler for pod deletions.
+func (manager *Manager) OnDeletePod(pod *slim_corev1.Pod) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	if len(manager.policyConfigs) == 0 {
+		return
+	}
+	id := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	podIPs := k8sUtils.ValidIPs(pod.Status)
+	if len(podIPs) == 0 {
+		return
+	}
+
+	var podData *podMetadata
+	var err error
+	if podData, err = manager.getPodMetadata(pod, podIPs); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sPodName:   pod.Name,
+			logfields.K8sNamespace: pod.Namespace,
+		}).Error("failed to get valid pod metadata")
```

Comment:
Let's have these msgs be a bit more descriptive where they failed because we are not returning an error to the caller so we don't really have a stack trace to understand what went wrong when we debug. My suggestion below is not that great, so you if have better wording then please go for it.

```suggestion
		}).Error("Failed to get valid pod metadata while cleaning up egress policy from pod")
```

Reply:
Fixed.